File: Android/TUIKit/TUICallKit/tuicallkit/src/main/java/com/tencent/qcloud/tuikit/tuicallkit/view/common/gridimage/GridImageData.java
package com.tencent.qcloud.tuikit.tuicallkit.view.common.gridimage;

import android.graphics.Bitmap;
import android.graphics.Color;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class GridImageData implements Cloneable {
    private static final int MAX_SIZE = 9;

    protected List<Object>         imageUrlList;
    protected Map<Integer, Bitmap> bitmapMap;
    protected int                  defaultImageResId;
    protected int                  bgColor = Color.parseColor("#cfd3d8");

    protected int targetImageSize;
    protected int maxWidth;
    protected int maxHeight;
    protected int rowCount;
    protected int columnCount;
    protected int gap = 6;

    public GridImageData() {
    }

    public GridImageData(List<Object> imageUrlList, int defaultImageResId) {
        this.imageUrlList = imageUrlList;
        this.defaultImageResId = defaultImageResId;
    }

    public int getDefaultImageResId() {
        return defaultImageResId;
    }

    public void setDefaultImageResId(int defaultImageResId) {
        this.defaultImageResId = defaultImageResId;
    }

    public List<Object> getImageUrlList() {
        return imageUrlList;
    }

    public void setImageUrlList(List<Object> imageUrlList) {
        this.imageUrlList = imageUrlList;
    }

    public void putBitmap(Bitmap bitmap, int position) {
        if (null != bitmapMap) {
            synchronized (bitmapMap) {
                bitmapMap.put(position, bitmap);
            }
        } else {
            bitmapMap = new HashMap<>();
            synchronized (bitmapMap) {
                bitmapMap.put(position, bitmap);
            }
        }
    }

    public Bitmap getBitmap(int position) {
        if (null != bitmapMap) {
            synchronized (bitmapMap) {
                return bitmapMap.get(position);
            }
        }
        return null;
    }

    public int size() {
        if (null != imageUrlList) {
            return Math.min(imageUrlList.size(), MAX_SIZE);
        } else {
            return 0;
        }
    }

    @Override
    protected GridImageData clone() throws CloneNotSupportedException {
        GridImageData gridImageData = (GridImageData) super.clone();
        if (imageUrlList != null) {
            gridImageData.imageUrlList = new ArrayList<>(imageUrlList.size());
            gridImageData.imageUrlList.addAll(imageUrlList);
        }
        if (bitmapMap != null) {
            gridImageData.bitmapMap = new HashMap<>();
            gridImageData.bitmapMap.putAll(bitmapMap);
        }
        return gridImageData;
    }
}

File: Android/TUIKit/TUICallKit/tuicallkit-kt/src/main/java/com/tencent/qcloud/tuikit/tuicallkit/view/common/gridimage/GridImageData.kt
package com.tencent.qcloud.tuikit.tuicallkit.view.common.gridimage

import android.graphics.Bitmap
import android.graphics.Color
import java.lang.Integer.min

class GridImageData : Cloneable {

    var imageUrlList: List<Any?>? = null
    var bitmapMap: MutableMap<Int, Bitmap?>? = null
    var defaultImageResId: Int = 0
    var bgColor: Int = Color.parseColor("#cfd3d8")
    var targetImageSize: Int = 0
    var maxWidth: Int = 0
    var maxHeight: Int = 0
    var rowCount: Int = 0
    var columnCount: Int = 0
    var gap: Int = 6

    constructor()

    constructor(imageUrlList: List<Any?>?, defaultImageResId: Int) {
        this.imageUrlList = imageUrlList
        this.defaultImageResId = defaultImageResId
    }

    fun putBitmap(bitmap: Bitmap, position: Int) {
        if (bitmapMap == null) {
            bitmapMap = HashMap()
        }
        bitmapMap?.let {
            synchronized(it) { it.put(position, bitmap) }
        }
    }

    fun getBitmap(position: Int): Bitmap? {
        bitmapMap?.let {
            synchronized(it) { return it[position] }
        }
        return null
    }

    fun size(): Int {
        imageUrlList?.let {
           return min(it.size, MAX_SIZE)
        }
        return 0
    }

    @Throws(CloneNotSupportedException::class)
    public override fun clone(): GridImageData {
        val gridImageData = super.clone() as GridImageData
        imageUrlList?.let {
            gridImageData.imageUrlList = ArrayList(it.size)
            (gridImageData.imageUrlList as ArrayList).addAll(it)
        }
        bitmapMap?.let {
            gridImageData.bitmapMap = HashMap()
            gridImageData.bitmapMap?.putAll(it)
        }
        return gridImageData
    }

    companion object {
        private const val MAX_SIZE = 9
    }
}